### DISCUSSION TITLE
LAN Modul nachrüsten für OpenDTU mit Standard ESP32 - Nötige Schritte?

### DISCUSSION THREAD

**Discussion by io2345** [tbnobody/OpenDTU#1727] [Q&A]
Hallo, habe einen ESP32 mit Funkmodul CMT2300A, auf dem OpenDTU läuft. 
Nun würde ich gerne noch ein LAN-Modul nachträglich "anflanschen" an den ESP32.
Kann mir jemand helfen und kurz beschreiben, wie das Schritt für Schritt zu machen wäre? 
Meiner Meinung nach könnte es entweder ein LAN8720 oder W5500 Modul sein (gibt's hier schon etwas zu beachten, oder ist egal, welches ich nehme?). Dieses müsste an die passenden PINs des ESP32 verbunden werden (welche Verbindungen sind nötig?) und am Ende müsste ggf. noch eine andere Firmware drauf (momentan v24.1.26) - oder man muss vielleicht auch nur (oder zusätzlich?) eine andere pinmappinging. json verwenden.
Vielleicht hat das ja schon jemand umgesetzt bei sich?
Geschwindigkeit der LAN-Verbindung ist übrigens egal, da läuft ja nix großartig drüber.

**Comment by D3R-ST3FAN** [comment]
Unterstützung für Ethernet gibt es doch schon sehr lange.
Es gibt ja auch bereits Profile für einige Boards Olimex 32 POE, WT32 ETH01, LilyGo TTGO ...
Die laufen alle mit den entsprechenden Pin Mappings. Hab ich selbst alle probiert.
Das Olimex ist ziemlich gut dokumentiert, da Open Source HW. Den Ethernet Teil (ist aber glaube ich ein 8710) kannst du theoretisch einfach anstecken. Ob sich das lohnt oder du lieber das Board wechselst musst du dir überlegen.

Wenn das Fusion Board mit der POE Extension mal im Hauptzweig unterstützt wird gibt es meiner Meinung nach die TOP Lösung.
(Die anderen oben gehen natürlich auch).

**Comment by io2345** [comment]
Danke für den Kommentar. Weiss schon, dass es Boards mit LAN-Anschluss gibt, aber mir geht es tatsächlich um die Erweiterung eines vorhandenen Standard ESP32-Wroom-Boards.
Als Module dachte ich an eines der beiden:
https://de.aliexpress.com/item/1005005992586738.html
https://de.aliexpress.com/item/1005006061643080.html
Mit "einfach anstecken" meinst du wahrscheinlich, die entsprechenden Pfostenstecker zu verbinden zwischen LAN-Modul und ESP-Board (?) - das war ja genau meine Frage, welche Verbindungen unbedingt nötig sind und welche Pins auf dem ESP32 für die Verarbeitung der Signale zuständig sind. 
Für das 8720-Modul hätte ich schon eine denkbare Belegung gefunden, siehe hier:
https://sautter.com/blog/ethernet-on-esp32-using-lan8720/
aber ich wollte natürlich am liebsten die gleichen PINs verwenden, die auch das Standard-OpenDTU-Release verwendet.

**Comment by D3R-ST3FAN** [comment]
[Hier](https://github.com/tbnobody/OpenDTU/tree/master/docs/DeviceProfiles) in den Profilen sind doch die GPIOs zu sehen die genutzt werden. Ansonsten kannst du ja auch dein eigenes Profil erzeugen. Das ist unabhängig von der Binary.

**Reply by io2345** [reply]
OK, verstehe. Man kann die GPIOs, die verwendet werden sollen, in der Pin_Mapping.json selbst festlegen. Das wusste ich bisher nicht.
Dann schauen wir mal, welche Werte wir benötigen. In allen Profilen mit "eth" Eintrag gibt es diese Werte:

enabled
phy_addr
power
mdc
mdio
type
clk_mode

Für enabled ist "true" klar.
phy_addr sollte (nach dem letzten Absatz im o.g. Sautter-Blog) "1" sein.
power liegt bei allen Profilen mit LAN auf GPIO12, ebenso mdc und mdio bei allen auf 23 bzw. 18.
type ist jeweils "0"
clk_mode ist schwieriger, da gibt es Profile mit 0 oder 3. Der jeweilige Wert könnte das "clock issue" Problem aus dem Sautter Blog adressieren. Hier hilft vermutlich nur ausprobieren.

Gut, damit ist für mich (in Verbindung mit dem Sautter-Blog) klar, dass ich an meinem OpenDTU-Modul GPIO 12 (da liegt bei mir bisher CLK vom CMT2300A-Modul) sowie zusätzlich die GPIOs 26 und 27 umlegen muss. Bisher ist GPIO26 der FCS und GPIO27 der CS vom Funkmodul. Diese GPIOs sind aber laut Blog fix für die Verwendung mit dem LAN8720-Modul. Weitere fixe GPIOs sind übrigens 22 (für TX1), 19 (für TX0), 21 (für TX_EN), 25 (für RX0) und 00 (für nINT/REFCLK).
OK, für 8720 sollte soweit alles klar sein.

Für das W5500-Modul habe ich dagegen noch keinen Plan, da sind die Bezeichnungen auch ganz anders. Hier gibt es:
MOSI
SCLK
SCNn
INTn
RSTn
MISO

Ich schreib hier mal weiter, wenn ich da was finde.

**Reply by tbnobody** [reply]
Bitte beachten, dass w5500 einen eigenen tcp Stack mitbringen und (noch) nicht unterstützt werden.

**Reply by io2345** [reply]
Ah, Danke für den Hinweis. Dann ist die Sache eh klar, wird also ein LAN8720-Modul.
